### DISCUSSION TITLE
Light: Move to Color with OnOff - Does this command exist?

### DISCUSSION THREAD

**Discussion by rechichidaniel** [Koenkk/zigbee2mqtt#16407] [General]
Hi,

What is the command/cluster for turning a light on with a chosen colour?
I can't find anything in [cluster.ts](https://github.com/Koenkk/zigbee-herdsman/blob/master/src/zcl/definition/cluster.ts)

When I attempt to turn the light on with a colour change from both Home Assistant and Zigbee2MQTT, I can see in my network traffic that the `Move to Color (0x07)` command on the `Color Control (0x0300)` cluster was sent, but the light didn't turn on.

Does Zigbee not support turning a light on with a colour setting?

Note that this does work for brightness where the `Move to Level with OnOff (0x04)` command on the `Level Control (0x0008)` cluster is used.

**Comment by rechichidaniel** [comment]
Changing to a Hue Bridge and looking at the traffic again shows that Hue does this in 2 stages. But this doesn't really work because it creates a super annoying lightning affect if the lights were set to white when they were turned off, and then turned back on with a different colour.

Hue first sends the `On (0x01)` command from the `On/Off (0x0006)` cluster, followed by `Move to Color (0x07)` from the `Color Control (0x0300)` cluster.

**Comment by donutVillan** [comment]
This should be possible with scenes as I understand it. 
I have a similar issue. I want to start the light with a certain color and a certain brightness (wake-up light). Starting the light at 100% brightness and then transition down to 1 isn’t particularly nice.
I haven’t tried this myself yet because I ran into an issue with zigbee2mqtt dropping a lot of my wall switches that required all of my attention but now it’s time to try and get this working.

**Comment by rechichidaniel** [comment]
Thanks DonutVillan. I'd really appreciate the help if you plan to look into this further.

I got as far as storing a scene using the Z2M Scene menu. But I don't know how to tell the light to change to the scene I stored.

When I click the recall button, Z2M shows a success message at the top right of the screen, but in my network traffic, my light responds with then `Unsupported Cluster Command (0x81)` status in the `Default Response (0x0b)` command from the `Scenes (0x0005)` cluster.

**Comment by rechichidaniel** [comment]
After a bit more playing around with scenes, wouldn't relying on scenes for this be a bit limiting because whatever you want to do has to be stored prior to use? In saying that though, I'd never tried to use device level scenes until just now and as you can see from my last message, I didn't get very far.

If scenes are the only or best way, then I'll start trying to figure them out.

**Reply by donutVillan** [reply]
As far as I understand they have to be stored prior to using them. I would have guessed that it's possible to do this automatically once when you start (re-start) home assistant (or whatever you are using). On my to-do list. Getting the light switches to work was more important for the WAF...

**Reply by rechichidaniel** [reply]
Hahaha WAF is important!! That's why I need these lights to turn on with a colour that's not blinding to someone that's trying to sleep...

**Reply by Ten0** [reply]
I'm trying to solve the same issue.
I also noticed that restoring scenes didn't have this wrong-color-turn-on issue.

It may be possible to create a scene directly: https://www.zigbee2mqtt.io/guide/usage/scenes.html#creating-a-scene

I have a tuya lightbulb that also doesn't respect transition=0 commands:
https://github.com/Koenkk/zigbee-herdsman-converters/blob/0aca991af6375b77fade002c9c2c1197eec8f21d/src/devices/tuya.ts#L298-L300
My thought was that we should be able to, instead of doing 0.1 when specified 0, create a scene with all the proper parameters set (we can reserve a high scene ID for this, it's probably fine for 99.999% of users), then immediately activate it.
However from my testing that message to create a scene ends up ignoring the `hue` parameter on my light (that is, it always turns on with a color that matches hue=0), despite it getting to the low level driver and actually creating the scene, respecting e.g. brightness:
```
zh:controller:endpoint: ZCL command 0xa4c13826f4d3f289/1 genScenes.add({"groupid":0,"sceneid":3,"scenename":"","transtime":3,"extensionfieldsets":[{"clstId":6,"len":1,"extField":[1]},{"clstId":8,"len":1,"extField":[128]},{"clstId":768,"len":4,"extField":[45350,19333]}]}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false})
```
and despite it being supposedly possible according to the lamp's documentation: https://developer.tuya.com/en/docs/connect-subdevices-to-gateways/tuya-zigbee-lighting-access-standard?id=K9ik6zvod83fi#title-14-DP6%20Scene and despite z2m thinking that it should be red when I recall the scene:
![image](https://github.com/user-attachments/assets/4fc8c5eb-820d-4ae4-baa5-ccaa2eff2ef4)

So overall I suspect a bug in how z2m builds the message that it should send to the lamp to create the scene with explicit color there, and it looks like once that bug is fixed we could consider using that create scene/activate scene trick to build a "move to color with on/off" command.

Unfortunately I do not know of a good way to test that theory: I could try to figure out what payload I should send to create the scene, but then to compare it with what currently happens I would need to:
1. send it to the device directly (however z2m does not have an interface for sending raw commands)
2. compare it with what z2m is actually sending (however the debug logs of `ember` don't print out the raw ZCL payload)

**Reply by rechichidaniel** [reply]
@Ten0 that sounds a lot like a bug I found with some 3A Nue lights, but I don't think it was caused by Z2M. I used a Zigbee sniffer to verify that the commands were being sent and received as intended.

Your lights may be different, but this is what I found with mine.

My steps:
1. Power cycle the light to get a fresh start.
2. Use any basic command that doesn't support a transition time ( `transition = 0` ) to verify that they work.
    * e.g. `On (0x01)` from the `On/Off (0x0006)` cluster.
3. Use any command that does support a transition time.
    * e.g. `Move to Level with OnOff (0x04)` command on the `Level Control (0x0008)` cluster.
5. Try a basic command again and notice that it won't work. The light will stay in the state it had after the transition command. However, the Zigbee sniffer showed that the light received the command and responded with an acknowledgement, choosing to ignore it internally.
    * e.g. set a colour. or `On (0x01)` from the `On/Off (0x0006)` cluster.
7. Power cycle the light and notice that it will have changed to reflect the commands it previously ignored, proving that the light did receive the commands, but chose to ignore them.

**Reply by Ten0** [reply]
Yes I agree that's probably a bug with the lamp itself, it's just that I noticed that scenes didn't have the issue so I figured we might be able to use that as a workaround instead.

**Comment by Danit2** [comment]
Do you know the command execute_if_off ? 
This is possible on the color control cluster and on the level control cluster. 

When this is active you can change the color or the brightness without to turn on the light. 
After this you can turn on the light with the rigt color and brightness. 

But this command is not working on every device.

**Reply by donutVillan** [reply]
Interesting! I have to look into this. Might be an easier way for me than using scenes.

**Reply by rechichidaniel** [reply]
Thanks. I'd never heard of `execute_if_off` before. Do you just add `"execute_if_off":true` to the payload?
I have my doubts about my lights supporting it when they get so many other commands wrong, but I think it's worth a try.

**Reply by Danit2** [reply]
no, you must first activate this with the command:
```
{"level_config": {"execute_if_off": true}}
```
and 
```
{"color_options": {"execute_if_off": true}}
```
then you can send the command likt this:
```
{"brightness": 179, "state": null}
```

But it can be, that the light lose this option when they lose the power. so after a power autage you must send the `execute_if_of` again.

**Reply by Danit2** [reply]
i use this for exempel for my bathroom light. 
At 22:00 i send a brightnes with 3 to the light. 
so when at night someone go in the bathroom and turn on the light with the local pushbutton the light will turn on at a minimum brightness. 
And in the morning i change it back to the brightness 179 that when someone go in the bathroom and need light that they will turn on with a good brightness for the day.

**Reply by rechichidaniel** [reply]
Thanks Danit2,
I tried but couldn't make it work because my light responds with `Status: Unsupported Attribute (0x86)`. But Wireshark can't decode that command properly so maybe I've done something wrong. Wireshark thinks I sent `Attribute: Unknown (0x000f)` from the `Write Attributes (0x02)` command on the `Level Control (0x0008)` cluster. It's the unknown attribute that's making me wonder.

Command json sent from a Node Red Inject node to a Zigbee2MQTT Out node.
```json
{
    "level_config": {
        "execute_if_off": true
    }
}
```
Coordinator Message
```javascript
Frame 2671: 50 bytes on wire (400 bits), 50 bytes captured (400 bits) on interface COM7, id 0
    (...)
IEEE 802.15.4 Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Network Layer Data, Dst: 0x1b6d, Src: 0x0000
    (...)
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
        .... ..00 = Frame Type: Data (0x0)
        .... 00.. = Delivery Mode: Unicast (0x0)
        ..0. .... = Security: False
        .0.. .... = Acknowledgement Request: False
        0... .... = Extended Header: False
    Destination Endpoint: 1
    Cluster: Level Control (0x0008)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 147
ZigBee Cluster Library Frame, Command: Write Attributes, Seq: 53
    Frame Control Field: Profile-wide (0x10)
        .... ..00 = Frame Type: Profile-wide (0x0)
        .... .0.. = Manufacturer Specific: False
        .... 0... = Direction: Client to Server
        ...1 .... = Disable Default Response: True
    Sequence Number: 53
    Command: Write Attributes (0x02)
    Attribute Field, Bitmap: 01
        Attribute: Unknown (0x000f)
        Data Type: 8-Bit Bitmap (0x18)
        Bitmap8: 0x01
```
Device Response
```javascript
Frame 2673: 49 bytes on wire (392 bits), 49 bytes captured (392 bits) on interface COM7, id 0
    (...)
IEEE 802.15.4 Data, Dst: 0x0000, Src: 0x1b6d
    (...)
ZigBee Network Layer Data, Dst: 0x0000, Src: 0x1b6d
    (...)
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
        .... ..00 = Frame Type: Data (0x0)
        .... 00.. = Delivery Mode: Unicast (0x0)
        ..0. .... = Security: False
        .0.. .... = Acknowledgement Request: False
        0... .... = Extended Header: False
    Destination Endpoint: 1
    Cluster: Level Control (0x0008)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 7
ZigBee Cluster Library Frame, Command: Write Attributes Response, Seq: 53
    Frame Control Field: Profile-wide (0x18)
        .... ..00 = Frame Type: Profile-wide (0x0)
        .... .0.. = Manufacturer Specific: False
        .... 1... = Direction: Server to Client
        ...1 .... = Disable Default Response: True
    Sequence Number: 53
    Command: Write Attributes Response (0x04)
    Status Record
        Status: Unsupported Attribute (0x86)
        Attribute: Unknown (0x000f)
```

**Reply by Danit2** [reply]
![image](https://user-images.githubusercontent.com/71522810/215759704-3609845f-1e68-48ca-9367-d00fc86d21b7.jpeg)

It can be that on your device this option is not working. You can test it also over the dev tool in the ui. 

There the execute_if_of is named "option".

**Reply by rechichidaniel** [reply]
Thanks Danit,

Unfortunately, these lights just don't seem to support it. But thanks anyway because this was definitely worth trying and it taught me something new. I can't seem to find the `execute_if_off` (or other spellings of it) in the Zigbee Cluster Library Specification, but it is mentioned in other Zigbee Alliance documents. I know it's not a command, but is it a standard attribute for other standard commands?

**Comment by rechichidaniel** [comment]
I'd like to keep looking for other ways to turn these lights on with different settings than they had when they were turned off. Does anyone have any other ideas?
